### PR TITLE
GPL-843 Added must_sequence and preferentially_sequence to BoxBuster.

### LINKED ISSUE #51: GPL-843 Flag 'Must sequence' and 'preferentially sequence' samples in box buster
**User story**
GPL-843 | As core LIMS we would like to enhance Boxbuster GUI to notify to operations a plate contains important samples, must sequence or preferentially sequence and it should be cherry picked preferentially

NB 
Must sequence - Business rules on Result, Controls, or CT values are ignored 
Preferentially Sequence - Flag as 'fit to pick' only if pass business rules

**Who are the primary contacts for this story**
Callum S (PAM)
Ryes (PAM)
Harry C
Philip VV
Eduardo MR

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] Show to operations via box buster (See mock up link below)
- [ ] Samples that must be sequenced, These are samples flagged True by PAM that must be sequenced regardless of probability will produce good sequence (Eg Sample bypass business rules on Result, If control or Ct values)  
- [ ] Samples that should be sequenced preferentially, These are samples flagged as True by PAM that *also* then  pass business rules on Result, Control and Ct filters

NB We are expecting preferentially sequence to be te most used field by PAM to prioritize samples. 

**Additional context**
See mock up: https://lucid.app/lucidchart/invitations/accept/af2a4c65-1a9a-498e-a77c-3dcc2e774e8f

### PR DESCRIPTION
**Pull request by harrietc52**
Used [mock up](https://lucid.app/lucidchart/invitations/accept/af2a4c65-1a9a-498e-a77c-3dcc2e774e8f)
Closes #51
Coupled with https://github.com/sanger/unified_warehouse/issues/172, https://github.com/sanger/lighthouse/issues/247

### CODE REVIEW

**Review by pjvv — COMMENTED** [review]
All good, couldn't help doing some gardening along the way so will need another thorough read through!

File: pages/reports.vue [hunk 1]
```diff
@@ -0,0 +1,153 @@
+<template>
```
Comment (harrietc52) [inline pages/reports.vue:1]: @pjvv Does index.vue want to be removed and add a redirect from root index to /reports?

Comment (pjvv) [reply pages/reports.vue:1]: Yeah, I'm going to try do that today.

Comment (pjvv) [reply pages/reports.vue:1]: I thought about leaving the page as a "Dashboard" we can build out with some nerdy stats as we go along. Boxes scanned, samples processed, files parsed etc.

File: test/pages/box_buster.spec.js [hunk 2]
```diff
@@ -64,227 +43,276 @@ describe('BoxBuster', () => {
...
-    wrapper = mount(BoxBuster, { localVue })
-    await wrapper.setData(data)
-    const row = wrapper
-      .find('table')
-      .findAll('tr')
-      .at(1)
-    expect(row.text()).toBe('AP-rna-1-8-posiYes8')
+    await wrapper.setData({ plates: [plateA] })
+    const row = wrapper.find('table').findAll('tr').at(1)
+    expect(row.text()).toContain(plateA.plate_barcode)
```
Comment (harrietc52) [inline test/pages/box_buster.spec.js:90]: If this test is checking plate map, I think it should be checking whether it has the string 'Yes' or 'No'

File: test/pages/box_buster.spec.js [hunk 3]
```diff
@@ -64,227 +43,276 @@ describe('BoxBuster', () => {
...
     const data = { plates: [plateC] }
-    wrapper = mount(BoxBuster, { localVue })
     await wrapper.setData(data)
-    const row = wrapper
-      .find('table')
-      .findAll('tr')
-      .at(1)
-    expect(row.text()).toBe('AP-rna-3-0-nmapNoN/A')
+    const row = wrapper.find('table').findAll('tr').at(1)
+    expect(row.text()).toContain(plateC.plate_barcode)
```
Comment (harrietc52) [inline test/pages/box_buster.spec.js:98]: Same here

File: test/pages/box_buster.spec.js [hunk 4]
```diff
@@ -64,227 +43,276 @@ describe('BoxBuster', () => {
...
     await flushPromises()
-    expect(wrapper.text()).toContain('Server Error')
+    expect(wrapper.find('#box-barcode').text()).toContain(ERROR_LABWHERE)
   })
 
-  it('falls back to a plate lookup if there appears to be no box', async () => {
-    getPlatesFromBoxBarcodes.mockResolvedValue({
-      success: false,
-      error: new Error('Server Error')
+  describe('#findBox', () => {
```
Comment (harrietc52) [inline test/pages/box_buster.spec.js:131]:
```suggestion
  describe('#platesProvider', () => {
```

File: test/pages/box_buster.spec.js [hunk 5]
```diff
@@ -64,227 +43,276 @@ describe('BoxBuster', () => {
...
-      'AP-rna-2-2-posi',
-      'AP-rna-3-0-nmap',
-      'AP-rna-4-5-posi',
-      'AP-rna-5-0-posi'
-    ]
-    getPlatesFromBoxBarcodes.mockResolvedValue({ success: true, barcodes })
-    lighthouse.findPlatesFromBarcodes.mockResolvedValue({
-      success: true,
-      plates: []
+  describe('#findPlates', () => {
```
Comment (harrietc52) [inline test/pages/box_buster.spec.js:190]: @emrojo Up to here :)

**Review by harrietc52 — COMMENTED** [review]
80% reviewed, @emrojo to continue tomorrow!
All good so far, minor comments

File: test/components/PrintLabelsRouter.spec.js [hunk 6]
```diff
@@ -19,15 +19,15 @@ describe('PrintLabelsRouter.vue', () => {
...
+      stubs: ['nuxt-link'],
     })
   })
 
   it('is a Vue instance', () => {
     expect(wrapper.findComponent(PrintLabelsRouter).exists()).toBeTruthy()
   })
 
-  each(links).it('will have a link to %s', (link) => {
+  each(links).test('will have a link to %s', (link) => {
```
Comment (emrojo) [inline test/components/PrintLabelsRouter.spec.js:30]: just of interest, what was the issue with it()?

**Review by emrojo — DISMISSED** [review]

File: .github/workflows/docker_dispatch.yml [hunk 7]
```diff
@@ -4,9 +4,9 @@ on:
   workflow_dispatch:
     inputs:
       image_tag:
-        description: "Image Tag"
+        description: 'Image Tag'
         required: true
-        default: "uat-..."
+        default: 'uat-...'
```
[the remark this answers is not in the document]
Comment (pjvv) [reply .github/workflows/docker_dispatch.yml:9]: It allows us to create images without having to create a release. From the actions tab you go to this workflow, enter the tag name and select the target branch and it creates an image from that target. Handy for creating images for UAT.

File: .gitignore [hunk 8]
```diff
@@ -88,3 +88,11 @@ sw.*
...
 # Vim swap files
 *.swp
+
+# Yarn berry
+.yarn/*
+!.yarn/releases
+!.yarn/plugins
+!.yarn/sdks
+!.yarn/versions
+.pnp.*
```
[the remark this answers is not in the document]
Comment (pjvv) [reply .gitignore:98]: The `!` negates a previous rule: https://git-scm.com/docs/gitignore#_pattern_format

File: README.md [hunk 9]
```diff
@@ -1,22 +1,95 @@
-# lighthouse-ui
+# Lighthouse UI
 
-> Basic website for Lighthouse
+![lint_and_test](https://github.com/sanger/lighthouse-ui/workflows/lint_and_test/badge.svg)
```
[the remark this answers is not in the document]
Comment (pjvv) [reply README.md:3]: Yeah, hoping so! :D

File: README.md [hunk 10]
```diff
@@ -1,22 +1,95 @@
...
-# install dependencies
-$ yarn install
+- Create and download reports which contain the fit to pick samples and their locations
+- The ability to scan the barcode of a box or plate and get information about it, such as:
+  + whether there is plate map data for the barcode(s)
+  + the number of fit to pick samples
+  + the number of 'must_sequence' samples
+  + the number of 'preferentially_sequence' samples
+- Create samples in Sequencescape when scanning in a box? (Sentinel)
+- Create a cherrypick batch from Sentinel?
```
[the remark this answers is not in the document]
Comment (pjvv) [reply README.md:16]: Good spot, will fill these in.

File: README.md [hunk 11]
```diff
@@ -1,22 +1,95 @@
...
+There are also some cypress tests which can be run as follows:
+
+    yarn run cypress run
+
+## Deployment
+
+This project uses a Docker image as the unit of deployment. To create a release for deployment, create a release
+in GitHub and wait for the GitHub action to create the Docker image.
+
+The release version should align with the [standards](./standards.md).
```
[the remark this answers is not in the document]
Comment (pjvv) [reply README.md:85]: Yes, thanks, will update.

File: cypress/integration/home_page.spec.js [hunk 12]
```diff
@@ -1,6 +1,6 @@
 describe('Index test', () => {
   it('Visits the Index', () => {
-    cy.visit('http://localhost:3000/')
-    cy.contains('Lighthouse positive samples reports')
+    cy.visit('http://localhost:3000/reports')
+    cy.contains('Lighthouse fit to pick samples reports')
```
[the remark this answers is not in the document]
Comment (pjvv) [reply cypress/integration/home_page.spec.js:4]: Thanks, have added it to the eslint config. Needs a little more work for some cypress tests but deferring that to later.

File: lefthook.yml [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    packages-audit:
+      tags: security
+      run: yarn audit
+
+pre-commit:
+  parallel: true
+  commands:
+    eslint:
+      glob: '*.{js,ts,vue}'
+      run: yarn lint {staged_files}
```
[the remark this answers is not in the document]
Comment (pjvv) [reply lefthook.yml:16]: Haven't added the credentials checking yet though.

File: pages/box_buster.vue [hunk 14]
```diff
@@ -118,54 +208,99 @@ export default {
...
     lighthouseFeedback() {
       if (this.lighthouseResponse.success === null) {
-        return 'Waiting for response from lighthouse...'
+        return 'Waiting for response from the lighthouse service...'
       } else {
         return extractError(this.lighthouseResponse)
       }
-    }
+    },
+    isError() {
```
[the remark this answers is not in the document]
Comment (pjvv) [reply pages/box_buster.vue:221]: Yes, as a `:show` binding on the alert box.

File: pages/box_buster.vue [hunk 15]
```diff
@@ -118,54 +208,99 @@ export default {
...
 
-      const response = await getPlatesFromBoxBarcodes(this.barcode)
-      this.labwhereResponse = response
-      if (response.success) {
-        this.findPlates(response)
-      } else {
-        // If it isn't a box, perhaps its a plate.
-        // Requirements were that we should allow plate lookups
-        this.findPlates({ barcodes: [this.barcode] })
+        const plates = await this.findPlatesOuter()
```
[the remark this answers is not in the document]
Comment (pjvv) [reply pages/box_buster.vue:242]: Refactored to better names.

File: pages/box_buster.vue [hunk 16]
```diff
@@ -118,54 +208,99 @@ export default {
...
+    refreshResults() {
+      this.provider()
+      if (this.$refs.plates_table) {
+        this.$refs.plates_table.refresh()
+      }
+    },
+    async provider() {
+      this.plates = await this.platesProvider()
+    },
+    plural_ize(word, count) {
```
[the remark this answers is not in the document]
Comment (pjvv) [reply pages/box_buster.vue:301]: Yeah, I couldn't think of another way but have found one! https://www.npmjs.com/package/vue-pluralize

File: pages/print_labels/ad_hoc_plate.vue [hunk 17]
```diff
@@ -19,29 +19,15 @@
...
         <p>
-          <label for="selectPrinter">
-            Which printer would you like to use?
-          </label>
-          <b-form-select
-            id="selectPrinter"
-            v-model="printer"
-            :options="printers"
-          ></b-form-select>
+          <label for="selectPrinter"> Which printer would you like to use? </label>
```
[the remark this answers is not in the document]
Comment (pjvv) [reply pages/print_labels/ad_hoc_plate.vue:22]: Yeah, something I also want to desperately refactor but thought this PR is heavy enough as it is! 😆

File: test/pages/box_buster.spec.js (hunk 2 above)
Comment (pjvv) [reply test/pages/box_buster.spec.js:90]: Good spot, have added extra test.

File: test/pages/box_buster.spec.js (hunk 3 above)
Comment (pjvv) [reply test/pages/box_buster.spec.js:98]: Also added.

File: test/components/PrintLabelsRouter.spec.js (hunk 6 above)
Comment (pjvv) [reply test/components/PrintLabelsRouter.spec.js:30]: Eslint was complaining about something so this was to appease it!

File: pages/box_buster.vue [hunk 18]
```diff
@@ -45,26 +45,30 @@
         <b-row>
           <b-col>
             <ul>
-              <li>Total of {{ total }} {{ plural_ize('plate', total) }} in box</li>
+              <li>Total of {{ total }} {{ 'plate' | pluralize(total) }} in the box</li>
```
Comment (harrietc52) [inline pages/box_buster.vue:48]: The pluralizing in this file is a bit confusing, using the `inclusive` flag, like from [here](https://www.npmjs.com/package/pluralize#usage) might help?
e.g `Total of {{ pluralize('plate', total, true) }} in the box`

File: plugins/vue-pluralize.js [hunk 19]
```diff
@@ -0,0 +1,4 @@
+import Vue from 'vue'
+import VuePluralize from 'vue-pluralize'
```
Comment (harrietc52) [inline plugins/vue-pluralize.js:2]: Just realised your not using the `pluralize` package, so maybe ignore above!

**Review by harrietc52 — COMMENTED** [review]
All good :) can't approve as I created the PR initially. @emrojo want to do the honours!?

**Pull request merged**